src/frontend/src/optimizer/plan_node/stream_locality_provider.rs

```diff
@@ -0,0 +1,204 @@
+            }
+        };
+
+        // LocalityProvider maintains the append-only behavior if input is append-only
+        let base = PlanBase::new_stream_with_core(
+            &core,
+            dist,
+            input.stream_kind(),
+            input.emit_on_window_close(),
+            input.watermark_columns().clone(),
+            input.columns_monotonicity().clone(),
+        );
+        StreamLocalityProvider { base, core }
+    }
+
+    pub fn locality_columns(&self) -> &[usize] {
+        &self.core.locality_columns
+    }
+}
+
+impl PlanTreeNodeUnary<Stream> for StreamLocalityProvider {
+    fn input(&self) -> PlanRef {
+        self.core.input.clone()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        let mut core = self.core.clone();
+        core.input = input;
+        Self::new(core)
+    }
+}
+
+impl_plan_tree_node_for_unary! { Stream, StreamLocalityProvider }
+impl_distill_by_unit!(StreamLocalityProvider, core, "StreamLocalityProvider");
+
+impl StreamNode for StreamLocalityProvider {
+    fn to_stream_prost_body(&self, state: &mut BuildFragmentGraphState) -> PbNodeBody {
+        let state_table = self.build_state_catalog(state);
+        let progress_table = self.build_progress_catalog(state);
+
+        let locality_provider_node = LocalityProviderNode {
+            locality_columns: self.locality_columns().iter().map(|&i| i as u32).collect(),
+            // State table for buffering input data
+            state_table: Some(state_table.to_prost()),
+            // Progress table for tracking backfill progress
+            progress_table: Some(progress_table.to_prost()),
+        };
+
+        PbNodeBody::LocalityProvider(Box::new(locality_provider_node))
+    }
+}
+
+impl ExprRewritable<Stream> for StreamLocalityProvider {
+    fn has_rewritable_expr(&self) -> bool {
+        false
+    }
+
+    fn rewrite_exprs(&self, _r: &mut dyn ExprRewriter) -> PlanRef {
+        self.clone().into()
+    }
+}
+
+impl ExprVisitable for StreamLocalityProvider {
+    fn visit_exprs(&self, _v: &mut dyn ExprVisitor) {
+        // No expressions to visit
+    }
+}
+
+impl StreamLocalityProvider {
+    /// Build the state table catalog for buffering input data
+    /// Schema: same as input schema (locality handled by primary key ordering)
+    /// Key: `locality_columns` (vnode handled internally by `StateTable`)
+    fn build_state_catalog(&self, state: &mut BuildFragmentGraphState) -> TableCatalog {
+        let mut catalog_builder = TableCatalogBuilder::default();
+        let input = self.input();
+        let input_schema = input.schema();
+
+        // Add all input columns in original order
+        for field in &input_schema.fields {
+            catalog_builder.add_column(field);
+        }
+
+        // Set locality columns as primary key.
+        for locality_col_idx in self.locality_columns() {
+            catalog_builder.add_order_column(*locality_col_idx, OrderType::ascending());
```

Comment:
Just realize that ideally we should also have `Order` as a requirement for "better locality".